datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaConnector.java

```diff
@@ -147,32 +156,48 @@ public synchronized void onAssignmentChange(List<DatastreamTask> tasks) {
-          stopTask(task, connectorTaskEntry);
-        });
+      if (toCancel.size() > 0) {
+        // Mark the connector task as stopped so that, in case stopping the task here fails for any reason in
+        // restartDeadTasks the task is not restarted
+        synchronized (_toStopTasks) {
+          toCancel.stream().forEach(task -> {
+            _toStopTasks.put(task, _runningTasks.get(task));
+            _runningTasks.remove(task);
+          });
+        }
+        stopUnassignedTasks();
       }
 
-      for (DatastreamTask task : tasks) {
-        ConnectorTaskEntry connectorTaskEntry = _runningTasks.get(task);
-        if (connectorTaskEntry != null) {
-          AbstractKafkaBasedConnectorTask kafkaBasedConnectorTask = connectorTaskEntry.getConnectorTask();
-          kafkaBasedConnectorTask.checkForUpdateTask(task);
-          // Make sure to replace the DatastreamTask with most up to date info
-          // This is necessary because DatastreamTaskImpl.hashCode() does not take into account all the
-          // fields/properties of the DatastreamTask (e.g. dependencies).
-          _runningTasks.remove(task);
-          _runningTasks.put(task, connectorTaskEntry);
-          continue; // already running
+      boolean toCallRestartDeadTasks = false;
+      // This is the only place where we synchronize on both _toStopTasks and _runningTasks. But if such a case
+      // arises elsewhere the order of synchronization should be _runningTasks and then _toStopTasks.
+      synchronized (_toStopTasks) {
+        for (DatastreamTask task : tasks) {
+          ConnectorTaskEntry connectorTaskEntry = _runningTasks.get(task);
+          if (connectorTaskEntry != null) {
+            AbstractKafkaBasedConnectorTask kafkaBasedConnectorTask = connectorTaskEntry.getConnectorTask();
+            kafkaBasedConnectorTask.checkForUpdateTask(task);
+            // Make sure to replace the DatastreamTask with most up to date info
+            // This is necessary because DatastreamTaskImpl.hashCode() does not take into account all the
+            // fields/properties of the DatastreamTask (e.g. dependencies).
+            _runningTasks.remove(task);
+            _runningTasks.put(task, connectorTaskEntry);
+            continue; // already running
+          } else {
+            if (_toStopTasks.containsKey(task)) {
+              toCallRestartDeadTasks = true;
+              connectorTaskEntry = _toStopTasks.remove(task);
+            } else {
+              connectorTaskEntry = createKafkaConnectorTask(task);
+            }
+            _runningTasks.put(task, connectorTaskEntry);
+          }
         }
-
-        _runningTasks.put(task, createKafkaConnectorTask(task));
+      }
+      // If any tasks pending stop were re-assigned to this host we explicitly call restartDeadTasks to ensure
+      // stop and (re)start of the task. We do this to not wait for next periodic invocation of restartDeadTasks.
```

Comment:
nit: reword: "We do this to avoid waiting for the next periodic invocation of restartDeadTasks."